base/inference.jl

```diff
@@ -1490,7 +1537,85 @@ end
 inlining_enabled() = (JLOptions().can_inline == 1)
 coverage_enabled() = (JLOptions().code_coverage != 0)
 
-function code_for_method(method::Method, atypes::ANY, sparams::SimpleVector, preexisting::Bool=false)
+# TODO: track the worlds for which this InferenceState
+# is being used, and split it if the WIP requires it?
+function converge_valid_age!(sv::InferenceState)
+    # push the validity range of sv into its fixedpoint callers
+    # recursing as needed to cover the graph
+    for (i, _) in sv.backedges
+        if i.fixedpoint
+            updated = false
+            if i.min_valid < sv.min_valid
+                i.min_valid = sv.min_valid
+                updated = true
+            end
+            if i.max_valid > sv.max_valid
+                i.max_valid = sv.max_valid
+                updated = true
+            end
+            @assert !isdefined(i.linfo, :def) || !i.cached || i.min_valid <= i.world <= i.max_valid "invalid age range update"
+            if updated
+                converge_valid_age!(i)
+            end
+        end
+    end
+    nothing
+end
+
+function update_valid_age!(min_valid::UInt, max_valid::UInt, sv::InferenceState)
+    sv.min_valid = max(sv.min_valid, min_valid)
+    sv.max_valid = min(sv.max_valid, max_valid)
+    @assert !isdefined(sv.linfo, :def) || !sv.cached || sv.min_valid <= sv.world <= sv.max_valid "invalid age range update"
+    nothing
+end
+update_valid_age!(edge::InferenceState, sv::InferenceState) = update_valid_age!(edge.min_valid, edge.max_valid, sv)
+update_valid_age!(li::MethodInstance, sv::InferenceState) = update_valid_age!(min_age(li), max_age(li), sv)
+
+function add_backedge(li::MethodInstance, caller::InferenceState)
+    # TODO: if li is not providing the inferred return type directly
+    #       the backedge must be to the WIP InferenceState, not the MethodInstance
+    isdefined(caller.linfo, :def) || return # don't add backedges to toplevel exprs
+    if caller.stmt_edges[caller.currpc] === ()
+        caller.stmt_edges[caller.currpc] = []
+    end
+    push!(caller.stmt_edges[caller.currpc], li)
```

Comment:
Should add a comment explaining that this is a temporary accumulation of information that will later be turned into an actual callee->caller backedge.